corehq/apps/reports/dispatcher.py

```diff
@@ -118,13 +128,30 @@ def _slug_alias(self, slug):
         return self.slug_aliases.get(slug)
 
     @datespan_default
-    def dispatch(self, request, domain=None, report_slug=None, render_as=None,
-                 permissions_check=None, *args, **kwargs):
-        render_as = render_as or 'view'
-        domain = domain or getattr(request, 'domain', None)
+    def dispatch(
+        self,
+        request,
+        domain: Optional[str] = None,
+        report_slug: Optional[str] = None,
+        render_as: Literal[
+            'async',
+            'deprecate',
+            'email',
+            'excel',
+            'export',
+            'filters',
+            'json',
+            'partial',
+            'print',
+            'view',
```

Comment:
Should this list be a constant or something? It feels funny to hard-code this in multiple places...

Reply:
I agree, it does feel wrong.

If I try to use a constant, I get 

> 'Literal' may be parameterized with literal ints, byte and unicode strings, bools, Enum values, None, other literal types, or type aliases to other literal types

I can go the other way around -- I can get the dictionary keys from the type definition:

```python
>>> ReportResponseType = Literal['async', 'deprecate', 'excel']  # etc.
>>> ReportResponseType.__args__
('async', 'deprecate', 'excel')
```

Zipping those up with the lambdas to make the dictionary isn't very readable. And what's worse, it's `Literal.__args__` in the `typing` library, and `Literal.__values__` in the `typing_extensions` library. So we definitely shouldn't be using it.

We could define the dict with its type, like ...

```python
    response_type_to_property: Dict[ReportResponseType, Callable] = {
        'async': lambda: report.async_response,
        'deprecate': lambda: report.deprecate_response,
        'excel': lambda: report.excel_response,
        # ...
    }
```
... but there is nothing preventing the dict and the type definition from getting out of sync, so I'm not sure this is helpful.

I'm open to suggestions, but I suspect that until `Literal` accepts constants, like you suggest, there might not be a nice way to DRY this.